File: vertx-lang/vertx-lang-jython/src/main/python_scripts/vertx.py
# Copyright 2011 the original author or authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

""" 
This module provides the entry point to the vert.x platform 
"""

import org.vertx.java.deploy.impl.VertxLocator
import org.vertx.java.core.json
import org.vertx.java.deploy.impl

from core.http import HttpServer, HttpClient
from core.net import NetServer, NetClient
from core.sock_js import SockJSServer
from core.handlers import TimerHandler, DoneHandler
from core.javautils import map_to_java, map_from_java

__author__ = "Scott Horn"
__email__ = "scott@hornmicro.com"
__credits__ = "Based entirely on work by Tim Fox http://tfox.org"

class Vertx(object):
    config = None

def create_http_server(**kwargs):
    """ Return a HttpServer """
    return HttpServer(**kwargs)

def create_http_client(**kwargs):
    """ Return a HttpClient """
    return HttpClient(**kwargs)

def create_net_server(**kwargs):
    """ Return a NetServer """
    return NetServer(**kwargs)

def create_net_client(**kwargs):
    """ Return a NetClient """
    return NetClient(**kwargs)

def create_sockjs_server(http_server):
    """ Return a SockJSServer """
    return SockJSServer(http_server)

def get_logger():
    """ Get the logger for the verticle """
    return org.vertx.java.deploy.impl.VertxLocator.container.getLogger()

def deploy_verticle(main, config=None, instances=1, handler=None):
    """Deploy a verticle. The actual deploy happens asynchronously

    Keyword arguments:
    @param main: the main of the verticle to deploy
    @param config: dict configuration for the verticle
    @param instances: number of instances to deploy
    @param handler: an handler that will be called when deploy has completed

    """
    if config != None:
        config = org.vertx.java.core.json.JsonObject(map_to_java(config))
  
    org.vertx.java.deploy.impl.VertxLocator.container.deployVerticle(main, config, instances, DoneHandler(handler))

def deploy_worker_verticle(main, config=None, instances=1, handler=None):
    """Deploy a worker verticle. The actual deploy happens asynchronously

    Keyword arguments:
    @param main: the main of the verticle to deploy
    @param config: dict configuration for the verticle
    @param instances: the number of instances to deploy
    @param handler: an handler that will be called when deploy has completed
    """
    if config != None:
        config = org.vertx.java.core.json.JsonObject(map_to_java(config))
    org.vertx.java.deploy.impl.VertxLocator.container.deployWorkerVerticle(main, config, instances, DoneHandler(handler))


def deploy_module(module_name, config=None, instances=1, handler=None):
    """Deploy a module. The actual deploy happens asynchronously

    Keyword arguments:
    @param module_name: The name of the module to deploy
    @param config: dict configuration for the module
    @param instances: Number of instances to deploy
    @param handler: an handler that will be called when deploy has completed
    """
    if config != None:
        config = org.vertx.java.core.json.JsonObject(map_to_java(config))
    org.vertx.java.deploy.impl.VertxLocator.container.deployModule(module_name, config, instances, DoneHandler(handler))

def undeploy_verticle(id, handler=None):
    """Undeploy a verticle

    Keyword arguments:
    @param id: the unique id of the deployment
    @param handler: an handler that will be called when undeploy has completed
    """
    org.vertx.java.deploy.impl.VertxLocator.container.undeployVerticle(id, DoneHandler(handler))

def undeploy_module(id, handler=None):
    """Undeploy a module

    Keyword arguments:
    @param id: the unique id of the module
    @param handler: an handler that will be called when undeploy has completed
    """
    org.vertx.java.deploy.impl.VertxLocator.container.undeployModule(id, DoneHandler(handler))

def config():
    """Get config for the verticle
    @return: dict config for the verticle
    """
    if Vertx.config is None:
        Vertx.config = map_from_java(org.vertx.java.deploy.impl.VertxLocator.container.getConfig().toMap())
    return Vertx.config

def java_vertx():
    return org.vertx.java.deploy.impl.VertxLocator.vertx

def set_timer(delay, handler):
    """Sets a one-shot timer that will fire after a certain delay.

    Keyword arguments:
    @param delay: the delay, in milliseconds
    @param handler: an handler that will be called when the timer fires
    @return: the unique id of the timer
    """
    return java_vertx().setTimer(delay, TimerHandler(handler))

def set_periodic(delay, handler):
    """Sets a periodic timer.

    Keyword arguments:
    @param delay: the period of the timer, in milliseconds
    @param handler: an handler that will be called each time the timer fires
    @return: the unique id of the timer
    """
    return java_vertx().setPeriodic(delay, TimerHandler(handler))

def cancel_timer(id):
    """Cancels a timer.

    Keyword arguments:
    @param id: the id of the timer, as returned from set_timer or set_periodic
    @return: true if the timer was cancelled, false if it wasn't found.
    """
    return java_vertx().cancelTimer(id)

def run_on_loop(handler):
    """Put the handler on the event queue for this loop so it will be run asynchronously
    ASAP after this event has been processed

    Keyword arguments:
    @param handler: an handler representing the code that will be run ASAP
    """
    java_vertx().runOnLoop(DoneHandler(handler))
 
def exit():
    """ Cause the container to exit """
    org.vertx.java.deploy.impl.VertxLocator.container.exit()

File: vertx-core/src/main/java/org/vertx/java/core/Vertx.java
/*
 * Copyright (c) 2011-2013 The original author or authors
 * ------------------------------------------------------
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * and Apache License v2.0 which accompanies this distribution.
 *
 *     The Eclipse Public License is available at
 *     http://www.eclipse.org/legal/epl-v10.html
 *
 *     The Apache License v2.0 is available at
 *     http://www.opensource.org/licenses/apache2.0.php
 *
 * You may elect to redistribute this code under either of these licenses.
 */

package org.vertx.java.core;

import org.vertx.java.core.datagram.DatagramSocket;
import org.vertx.java.core.datagram.InternetProtocolFamily;
import org.vertx.java.core.dns.DnsClient;
import org.vertx.java.core.eventbus.EventBus;
import org.vertx.java.core.file.FileSystem;
import org.vertx.java.core.http.HttpClient;
import org.vertx.java.core.http.HttpServer;
import org.vertx.java.core.net.NetClient;
import org.vertx.java.core.net.NetServer;
import org.vertx.java.core.shareddata.SharedData;
import org.vertx.java.core.sockjs.SockJSServer;

import java.net.InetSocketAddress;

/**
 * The control centre of the Vert.x Core API.<p>
 * You should normally only use a single instance of this class throughout your application. If you are running in the
 * Vert.x container an instance will be provided to you.<p>
 * If you are using Vert.x embedded, you can create an instance using one of the static {@code VertxFactory.newVertx}
 * methods.<p>
 * This class acts as a factory for TCP/SSL and HTTP/HTTPS servers and clients, SockJS servers, and provides an
 * instance of the event bus, file system and shared data classes, as well as methods for setting and cancelling
 * timers.<p>
 * Instances of this class are thread-safe.<p>
 *
 * @author <a href="http://tfox.org">Tim Fox</a>
 */
public interface Vertx {

  /**
   * Create a TCP/SSL server
   */
  NetServer createNetServer();

  /**
   * Create a TCP/SSL client
   */
  NetClient createNetClient();

  /**
   * Create an HTTP/HTTPS server
   */
  HttpServer createHttpServer();

  /**
   * Create a HTTP/HTTPS client
   */
  HttpClient createHttpClient();

  /**
   * Create a new {@link DatagramSocket}.
   *
   * @param family  use {@link InternetProtocolFamily} to use for multicast. If {@code null} is used it's up to the
   *                operation system to detect it's default.
   * @return socket the created {@link DatagramSocket}.
   */
  DatagramSocket createDatagramSocket(InternetProtocolFamily family);

  /**
   * Create a SockJS server that wraps an HTTP server
   */
  SockJSServer createSockJSServer(HttpServer httpServer);

  /**
   * The File system object
   */
  FileSystem fileSystem();

  /**
   * The event bus
   */
  EventBus eventBus();

  /**
   * Return the {@link DnsClient}
   */
  DnsClient createDnsClient(InetSocketAddress... dnsServers);

  /**
   * The shared data object
   */
  SharedData sharedData();

  /**
   * Set a one-shot timer to fire after {@code delay} milliseconds, at which point {@code handler} will be called with
   * the id of the timer.
   * @return the unique ID of the timer
   */
  long setTimer(long delay, Handler<Long> handler);

  /**
   * Set a periodic timer to fire every {@code delay} milliseconds, at which point {@code handler} will be called with
   * the id of the timer.
   * @return the unique ID of the timer
   */
  long setPeriodic(long delay, Handler<Long> handler);

  /**
   * Cancel the timer with the specified {@code id}. Returns {@code} true if the timer was successfully cancelled, or
   * {@code false} if the timer does not exist.
   */
  boolean cancelTimer(long id);

  /**
   * @return The current context
   */
  Context currentContext();

  /**
   * Put the handler on the event queue for the current loop (or worker context) so it will be run asynchronously ASAP after this event has
   * been processed
   */
  void runOnContext(Handler<Void> action);

  /**
   * Is the current thread an event loop thread?
   * @return true if current thread is an event loop thread
   */
  boolean isEventLoop();

  /**
   * Is the current thread an worker thread?
   * @return true if current thread is an worker thread
   */
  boolean isWorker();

  /**
	 * Stop the eventbus and any resource managed by the eventbus.
	 */
	void stop();
}
